File: app/jobs/email_enrichment_job.rb
class EmailEnrichmentJob < ApplicationJob
  queue_as :default

  retry_on StandardError, wait: :exponentially_longer, attempts: 2 do |job, exception|
    contact_id = job.arguments.first
    contact = Contact.find_by(id: contact_id)
    Rails.logger.warn("Email enrichment failed for contact #{contact_id}: #{exception.message}")
  end

  discard_on ActiveRecord::RecordNotFound do |job, exception|
    Rails.logger.error("Contact not found for email enrichment: #{exception.message}")
  end

  # Note: This job expects a contact_id (Integer), not a Contact object
  # This follows Sidekiq best practice of passing IDs to avoid serialization issues
  def perform(contact_id)
    contact = Contact.find(contact_id)
    # Skip if already enriched
    if contact.email_enriched?
      Rails.logger.info("Skipping contact #{contact.id}: email already enriched")
      return
    end

    # Skip if email enrichment disabled
    credentials = TwilioCredential.current
    unless credentials&.enable_email_enrichment
      Rails.logger.info("Email enrichment disabled in settings")
      return
    end

    # Perform enrichment
    Rails.logger.info("Enriching email data for contact #{contact.id}")

    success = EmailEnrichmentService.enrich(contact)

    if success
      Rails.logger.info("Successfully enriched contact #{contact.id} with email data")

      # Queue duplicate detection after email enrichment
      if credentials&.enable_duplicate_detection
        DuplicateDetectionJob.perform_later(contact_id)
      end
    else
      Rails.logger.info("No email data found for contact #{contact.id}")
    end

  rescue StandardError => e
    Rails.logger.error("Unexpected error enriching email for contact #{contact.id}: #{e.class} - #{e.message}")
    raise
  end
end

File: laravel-app/app/Jobs/EmailEnrichmentJob.php
<?php

namespace App\Jobs;

use App\Models\Contact;
use App\Models\TwilioCredential;
use App\Services\EmailEnrichmentService;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Log;
use Exception;

class EmailEnrichmentJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $contactId;

    /**
     * The number of times the job may be attempted.
     *
     * @var int
     */
    public $tries = 2;

    /**
     * The number of seconds to wait before retrying.
     *
     * @var int
     */
    public $backoff = [10, 30];

    /**
     * The maximum number of seconds the job can run before timing out.
     *
     * @var int
     */
    public $timeout = 60;

    /**
     * Create a new job instance.
     *
     * @param int $contactId
     */
    public function __construct(int $contactId)
    {
        $this->contactId = $contactId;
        $this->onQueue('default');
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        try {
            $contact = Contact::findOrFail($this->contactId);

            // Skip if already enriched
            if ($contact->email_enriched) {
                Log::info("Skipping contact {$contact->id}: email already enriched");
                return;
            }

            // Skip if email enrichment disabled
            $credentials = TwilioCredential::current();
            if (!$credentials || !$credentials->enable_email_enrichment) {
                Log::info("Email enrichment disabled in settings");
                return;
            }

            // Perform enrichment
            Log::info("Enriching email data for contact {$contact->id}");

            $service = new EmailEnrichmentService();
            $success = $service->enrich($contact);

            if ($success) {
                Log::info("Successfully enriched contact {$contact->id} with email data");

                // Queue duplicate detection after email enrichment
                if ($credentials->enable_duplicate_detection) {
                    DuplicateDetectionJob::dispatch($this->contactId);
                }
            } else {
                Log::info("No email data found for contact {$contact->id}");
            }

        } catch (Exception $e) {
            Log::error("Unexpected error enriching email for contact {$this->contactId}: " . get_class($e) . " - {$e->getMessage()}");
            throw $e;
        }
    }

    /**
     * Handle a job failure.
     *
     * @param  \Throwable  $exception
     * @return void
     */
    public function failed(\Throwable $exception)
    {
        Log::warning("Email enrichment failed for contact {$this->contactId}: {$exception->getMessage()}");
    }
}
